### ISSUE TITLE
model

### ISSUE THREAD

**Issue by linhanxiao** [junyanz/CycleGAN#54]

**Comment by linhanxiao** [comment]
I want to add regularization item to the entropyloss，should I directly add lamda×penal to the entropyloss in torch?，or should I create new Criterion Class inherited from nn.Criterion? My
regularization is not L2 regularization or dropout

**Comment by junyanz** [comment]
I think you can directly add the lambda.

**Comment by linhanxiao** [comment]
my regularization equation is totalloss=entropyloss+lamda∗‖AA^T −I‖^2,the A is one network layer output, so in my new class inherited from CrossEntropyCriterion,I only need to change updateOutput method by adding lamda*||AA^T-l||^2 to self.output ,is that right?

**Comment by linhanxiao** [comment]
i only change updateOuput method，how does the regulariztion information in totalloss backward

**Comment by linhanxiao** [comment]
i am from china,Can i speak with you in chinese

**Comment by linhanxiao** [comment]
Are you mean that i direct add lamda*penal item to the CrossEntropyCriterion.output?

**Comment by linhanxiao** [comment]
i only change updateOuput method，how does the regulariztion information in totalloss backward

**Comment by junyanz** [comment]
You don't have to implement a new Criterion Class. You can just calculate your loss in the fGx_basic and fDx_basic

**Comment by linhanxiao** [comment]
Are you mean that after the NetD and CriterionGan forward methods,i should add lamda*penal item to Criterion loss,then execute
the backward methods

**Comment by junyanz** [comment]
I think you can try it. You can also try pytorch which might be easier to implement stuff.

**Comment by linhanxiao** [comment]
the pytorch support the autograd,the torch does not support the autograd,so my question is why the implementation of model regularization in  torch will be similar to the implementation of model regularization in pytorch

**Comment by linhanxiao** [comment]
@junyanz ,Can I apply cyclegan to the problem of image caption?

**Comment by linhanxiao** [comment]
@junyanz the following is my loss regularization code,I directly added the penal item to crit.output,is it right for model regularization?


                                  **net:forward(sample.input)
                                  crit:forward(net.output, sample.target)
                                    local A
                                    for _,b in ipairs(_G.model.selfattentivesoftmax) do
                                        A=b.output    
                                    end
                                    local B=A:clone()
                                   for i=1,B:size(1) do
                                    A=B[i]:clone()
                                    local AAT=torch.mm(A,A:t())
                                    local I=torch.eye(A:size(1))
                                    local P=torch.norm( AAT - I, 2 )
                                    local penal=P*P
                                    penal = penal/A:size(2)
                                    crit.output=crit.output+_G.model.selfattentivelamda*penal
                                   end
                                    crit:backward(net.output, sample.target)
                                    net:backward(sample.input, crit.gradInput)**

**Comment by linhanxiao** [comment]
@junyanz  the following is my loss regularization code,I directly added the penal  item to crit.output,is it right for model regularization?                              
 net:forward(sample.input)                               
crit:forward(net.output, sample.target)                                 
local A                              
 for _,b in ipairs(_G.model.selfattentivesoftmax) do                                 
   A=b.output                                     
end                                 
local B=A:clone()                                
for i=1,B:size(1) do                                 
A=B[i]:clone()                                 
local AAT=torch.mm(A,A:t())                                 
local I=torch.eye(A:size(1))                                 
local P=torch.norm( AAT - I, 2 )                                 
local penal=P*P                                 
penal = penal/A:size(2)                                 
crit.output=crit.output+_G.model.selfattentivelamda*penal                                
end                                 
crit:backward(net.output, sample.target)                                 
net:backward(sample.input, crit.gradInput)

**Comment by linhanxiao** [comment]
@junyanz the following is my loss regularization code,I directly added the penal item to crit.output,is it right for model regularization?
net:forward(sample.input)
crit:forward(net.output, sample.target)
local A
for _,b in ipairs(_G.model.selfattentivesoftmax) do
    A=b.output
end
local B=A:clone()
for i=1,B:size(1) do
   A=B[i]:clone()
   local AAT=torch.mm(A,A:t())
   local I=torch.eye(A:size(1))
   local P=torch.norm( AAT - I, 2 )
   local penal=PP
   penal = penal/A:size(2)
   crit.output=crit.output+_G.model.selfattentivelamd*apenal
end
crit:backward(net.output, sample.target)
net:backward(sample.input, crit.gradInput)

**Comment by linhanxiao** [comment]
@junyanz the following is my loss regularization code,I directly added the penal item to crit.output,is it right for model regularization?
net:forward(sample.input)
crit:forward(net.output, sample.target)
local A
for _,b in ipairs(_G.model.selfattentivesoftmax) do
A=b.output
end
local B=A:clone()
for i=1,B:size(1) do
A=B[i]:clone()
local AAT=torch.mm(A,A:t())
local I=torch.eye(A:size(1))
local P=torch.norm( AAT - I, 2 )
local penal=P*P
penal = penal/A:size(2)
crit.output=crit.output+_G.model.selfattentivelamd*apenal
end
crit:backward(net.output, sample.target)
net:backward(sample.input, crit.gradInput)

**Comment by linhanxiao** [comment]
@junyanz the following is my loss regularization code,I directly added the penal item to crit.output,is it right for model regularization?
net:forward(sample.input)
crit:forward(net.output, sample.target)
local A
for _,b in ipairs(_G.model.selfattentivesoftmax) do
    A=b.output
end
local B=A:clone()
for i=1,B:size(1) do
   A=B[i]:clone()
   local AAT=torch.mm(A,A:t())
   local I=torch.eye(A:size(1))
   local P=torch.norm( AAT - I, 2 )
   local penal=P**P
   penal = penal/A:size(2)
   crit.output=crit.output+_G.model.selfattentivelamda**penal
end
crit:backward(net.output, sample.target)
net:backward(sample.input, crit.gradInput)

**Comment by linhanxiao** [comment]
@junyanz the following is my loss regularization code,I directly added the penal item to crit.output,is it right for model regularization?
net:forward(sample.input)
crit:forward(net.output, sample.target)
local A
for _,b in ipairs(_G.model.selfattentivesoftmax) do
    A=b.output
end
local B=A:clone()
for i=1,B:size(1) do
   A=B[i]:clone()
   local AAT=torch.mm(A,A:t())
   local I=torch.eye(A:size(1))
   local P=torch.norm( AAT - I, 2 )
   local penal=P*P
   penal = penal/A:size(2)
   crit.output=crit.output+_G.model.selfattentivelamda*penal
end
crit:backward(net.output, sample.target)
net:backward(sample.input, crit.gradInput)
